add components for surface snow data

emilyhcliu opened the pull request:
This PR includes the following changes:
1. Add mapping and Python for surface snow data
2. Move test configuration from /spoc/dump/config to /spoc/ush/test
3. Remove an old test configuration file
4. Update README for test (/spoc/ush/test)
5. Change filename convention change from `bufr2ioda` to `bufr` (more general)
6. Fix Python coding norm errors  (ctest is adding to obsForge - in progress [obsForge PR #4](https://github.com/NOAA-EMC/obsForge/pull/4)

emilyhcliu committed:
Remove old file; fix python coding norms

emilyhcliu committed:
Add mapping and python for sfcsno

emilyhcliu committed:
Add sfcsno test configuration files and move conf from dump to test

emilyhcliu committed:
Update README

nicholasesposito commented on dump/mapping/bufr_sfcsno_mapping.yaml:
```diff
@@ -0,0 +1,79 @@
+      coordinates: "longitude latitude"
+      source: variables/stationIdentification
+      longName: "Identification of Observing Location"
+      units: "index"
```
What is "index" ? Should this always be used for stationIdentification?

rmclaren commented on dump/mapping/bufr_sfcsno.py:
```diff
@@ -0,0 +1,189 @@
+    container.replace('variables/totalSnowDepth', snod)
+    snod_upd = container.get('variables/totalSnowDepth')
+
+    masked_container = _mask_container(container, (~snod.mask))
```
Instead of making those **snod** values equal 0, did you mean to mask them out instead? In this case you can modify the snod.mask directly (the True value means to mask out for numpy masked arrays).

emilyhcliu replied:
@rmclaren I took the [sfcsno Python code](https://github.com/NOAA-EMC/GDASApp/blob/91f160e614f8abef21529a9611f8aab6105f0ab3/ush/ioda/bufr2ioda/bufr2ioda_sfcsno_bufr_encoder.py) from GDASApp [PR #1368](https://github.com/NOAA-EMC/GDASApp/pull/1368) and modified it to following our Python code template.  I did not change how the derived variable was done.  For sfcsno converter, the data is read out from BUFR (10K+ data) and then applies data filtering to remove missing data.  The output file ends up with much less data (~3K).

nicholasesposito commented on ush/test/config/bufr_script_backend_sfcsno.yaml:
```diff
@@ -0,0 +1,22 @@
+time window:
+  begin: "2018-04-14T21:00:00Z"
+  end: "2023-12-15T03:00:00Z"
```
`  bound to include: begin
`

emilyhcliu replied:
fixed.

nicholasesposito commented on ush/test/config/bufr_bufr_backend_sfcsno.yaml:
```diff
@@ -0,0 +1,20 @@
+time window:
+  begin: "2018-04-14T21:00:00Z"
+  end: "2023-12-15T03:00:00Z"
```
`  bound to include: begin
`

emilyhcliu replied:
Fixed!

rmclaren commented on dump/mapping/bufr_sfcsno.py:
```diff
@@ -0,0 +1,189 @@
+logger = Logger('BUFR2IODA_sfcsno.py', level=log_level, colored_log=False)
+
+
+def logging(comm, level, message):
```
We might want to make a logging module that we can import, so that we do not need to redefine this function in every **py** mapping file. 

We don't have to do this now.

emilyhcliu replied:
I will do this in my next PR.

rmclaren commented on ush/test/README.md:
```diff
@@ -1,5 +1,5 @@
 ## Test bufr_query mapping, Python converter script, and ioda configuration YAML in obsForge
-This is a prototype for testing BUFR to IODA conversion and is still evolving.
+This is a prototype for testing BUFR to NETCDF/IODA conversion and is still evolving.
```
We will use this for ZARR as well... These could also be used in IODA backends too.

rmclaren commented on ush/test/README.md:
```diff
@@ -33,20 +29,22 @@ This is a prototype for testing BUFR to IODA conversion and is still evolving.
+- Required input files in ./work_dir:
   
-   - bufr_satwnd_amv_goes.py
+   - bufr_satwnd_amv_goes.py (copied /spoc/dump/mapping)
```
why copying things?

emilyhcliu replied:
This is a stand-alone test.  So, the uses need to gather (copy) the required input files (bufr, mapping, python script, configuration files)  in their working directory along with the test script (encodeBufr.sh) to run the test.

rmclaren commented on ush/test/config/bufr_bufr_backend_sfcsno.yaml:
bufr_bufr_backend ? maybe rename

emilyhcliu replied:
We have four tests:
bufr2netcdf
script2netcdf
bufr_backend
script_backend

How about we change the two backend tests to the following:
bufr_backend ---> bufr4backend
script_backend ---> script4backend

rmclaren replied:
sounds good.

PraveenKumar-NOAA commented on dump/mapping/bufr_sfcsno_mapping.yaml:
```diff
@@ -0,0 +1,79 @@
+    - name: "MetaData/stationElevation"
+      coordinates: "longitude latitude"
+      source: variables/stationElevation
+      longName: "Height of Station"
```
Unit is missing for the stationElevation!

PraveenKumar-NOAA commented on dump/mapping/bufr_sfcsno_mapping.yaml:
```diff
@@ -0,0 +1,79 @@
+      coordinates: "longitude latitude"
+      source: variables/totalSnowDepth
+      longName: "Total Snow Depth"
+      units: "mm"
```
Change units 'mm' to 'm'.

emilyhcliu replied:
Good Catch

emilyhcliu committed:
Add "bound to include" for time window

emilyhcliu committed:
rename bufr_backend to bufr4backend and script_backend to script4backend

emilyhcliu committed:
rename bufr_backend to bufr4backend and script_backend to script4backend

emilyhcliu committed:
fix typo

emilyhcliu committed:
Update readme

emilyhcliu committed:
change output directory (e.g. bufr_backend to bufr4backend)

emilyhcliu commented:
This is ready for merge.  I need one approval.  Thanks.

PraveenKumar-NOAA reviewed (dismissed):
Looks good!

emilyhcliu committed:
Add data receipt time

emilyhcliu commented:
@PraveenKumar-NOAA @nicholasesposito @rmclaren 
I updated the scripts/yamls, and readme documentation based on your review comments:
- changed totalSnowDepth unit from `mm` to `m` following IODA data convention
- added unit `m` to stationEvelation
- the unit for `stationIdentification` is set to `Index` for now since there is no unit defined in IODA data convention for `stationIdentification`.  
- added `bound to include:begin" for time window in the test configuration YAMLs
- updated README file for testing

I also updated the following:
- added observation receipt time (dataReceiptTime) to YAML for snow and satwind data
- updated test reference files (uploaded to EMCRZDM)

Please review them one more time and approve these updates if they look good to you.

There will be no more updates from me for this PR.  I will merge the PR as soon as I get your approvals.  Thanks!

nicholasesposito reviewed (approved):
The rest looks good to me!
